src/osp/Active/SysPhysics.cpp

```diff
@@ -0,0 +1,281 @@
+
+    // Perform first-time initialization of rigidbody ancestor component
+    if (pCompAncestor == nullptr)
+    {
+        auto [ent, pCompBody] = find_rigidbody_ancestor(rScene, childEntity);
+        if (pCompBody == nullptr)
+        {
+            // childEntity has no rigidbody ancestor
+            return nullptr;
+        }
+
+        rReg.emplace<ACompRigidbodyAncestor>(childEntity);
+    }
+    // childEntity now has an ACompRigidbodyAncestor
+
+    auto& rRbAncestor = rReg.get<ACompRigidbodyAncestor>(childEntity);
+    ActiveEnt& rAncestor = rRbAncestor.m_ancestor;
+
+    // Ancestor entity already valid
+    if (rReg.valid(rAncestor))
+    {
+        return &rRbAncestor;
+    }
+
+    // Rigidbody ancestor not set yet
+    auto [bodyEnt, compBody] = find_rigidbody_ancestor(rScene, childEntity);
+
+    if (compBody == nullptr)
+    {
+        SPDLOG_LOGGER_WARN(rScene.get_application().get_logger(),
+                          "No rigid body!");
+        return nullptr;
+    }
+
+    // Initialize ACompRigidbodyAncestor
+    rAncestor = bodyEnt;
+    rRbAncestor.m_relTransform =
+        find_transform_rel_rigidbody_ancestor(rScene, childEntity);
+    //TODO: this transformation may change and need recalculating
+
+    return &rRbAncestor;
+}
+
+/* Since masses are usually stored in the rigidbody's children instead of the
+ * rigidbody root, the function provides the option to ignore the root entity's
+ * mass (this is the default). Otherwise, the root of each call will be double
+ * counted since it will also be included in both the loop and the root of the
+ * next call. This makes it possible for a rigidbody to have a convenient
+ * root-level mass component set by some external system without interfering
+ * with physics calculations that depend on mass components to determine the
+ * mechanical behaviors of the rigidbody.
+ */
+template <SysPhysics::EIncludeRootMass INCLUDE_ROOT_MASS>
+Vector4 SysPhysics::compute_hier_CoM(ActiveScene& rScene, ActiveEnt root)
+{
+    Vector3 localCoM{0.0f};
+    float localMass = 0.0f;
+
+    /* Include the root entity's mass. Skipped by default to avoid
+     * double-counting in recursion, and because some subhierarchies
+     * may have a total mass stored at their root for easy external access
+     */
+    if constexpr (INCLUDE_ROOT_MASS)
+    {
+        auto const* rootMass = rScene.reg_try_get<ACompMass>(root);
+        if (rootMass != nullptr) { localMass += rootMass->m_mass; }
+    }
+
+    for (ActiveEnt nextChild = rScene.reg_get<ACompHierarchy>(root).m_childFirst;
+        nextChild != entt::null;)
+    {
+        auto const &childHier = rScene.reg_get<ACompHierarchy>(nextChild);
+        auto const *childTransform = rScene.get_registry().try_get<ACompTransform>(nextChild);
+        auto const *massComp = rScene.get_registry().try_get<ACompMass>(nextChild);
+        Matrix4 childMatrix{Magnum::Math::IdentityInit};
```

Comment:
Minor, but

```
Matrix4 childMatrix = chioldTransform != nullptr
                                  ? childTransform->m_transform
                                  : Magnum::Math::IdentityInit;
```
should avoid double initializing the childMatrix variable.

Could be the compiler is smart enough to do this regardless, using control flow analysis, but it's never obvious when the compiler is going to actually be that smart.